xdem/terrain.py

```diff
@@ -230,3 +259,156 @@ def hillshade(
+        >>> dem = np.array([[1, 1, 1],
+        ...                 [1, 2, 1],
+        ...                 [1, 1, 1]], dtype="float32")
+        >>> coeffs = get_quadric_coefficients(dem, resolution=1.0)
+        >>> coeffs.shape
+        (9, 3, 3)
+        >>> coeffs[:, 1, 1]
+        array([ 1.,  0.,  0., -1., -1.,  0.,  0.,  0.,  2.])
+
+    :returns: An array of coefficients for each pixel of shape (9, row, col).
+    """
+    # This function only formats and validates the inputs. For the true functionality, see _get_quadric_coefficients()
+    dem_arr = xdem.spatial_tools.get_array_and_mask(dem)[0]
+
+    if len(dem_arr.shape) != 2:
+        raise ValueError(
+            f"Invalid input array shape: {dem.shape}, parsed into {dem_arr.shape}. "
+            "Expected 2D array or 3D array of shape (1, row, col)"
+        )
+
+    if any(dim < 3 for dim in dem_arr.shape):
+        raise ValueError(f"DEM (shape: {dem.shape}) is too small. Smallest supported shape is (3, 3)")
+
+    # Resolution is in other tools accepted as a tuple. Here, it must be just one number, so it's best to sanity check.
+    if isinstance(resolution, Sized):
+        raise ValueError("Resolution must be the same for X and Y directions")
+
+    # Try to run the numba JIT code. It should never fail at this point, so if it does, it should be reported!
+    try:
+        coeffs = _get_quadric_coefficients(dem_arr, resolution)
+    except Exception as exception:
+        raise RuntimeError("Unhandled numba exception. Please raise an issue of what happened.") from exception
+
+    return _get_quadric_coefficients(dem_arr, resolution)
+
+
+@numba.njit(parallel=True)
+def _get_quadric_coefficients(
+    dem: np.ndarray,
+    resolution: float,
+) -> np.ndarray:
+    """
+    Run the pixel-wise analysis in parallel.
+
+    See the xdem.terrain.get_quadric_coefficients() docstring for more info.
+    """
+    # Rename the resolution to be consistent with the ArcGIS reference.
+    L = resolution
+
+    # Allocate the output.
+    output = np.empty((9,) + dem.shape, dtype=dem.dtype) + np.nan
+
+    # Loop over every pixel concurrently.
+    for i in numba.prange(dem.size):
+        # Derive its associated row and column index.
+        col = i % dem.shape[1]
+        row = int(i / dem.shape[1])
+
+        # Extract the pixel and its 8 immediate neighbours.
+        # If the border is reached, just duplicate the closest neighbour to obtain 9 values.
+        Z = np.empty((9,), dtype=dem.dtype)
+        count = 0
+        for j in range(-1, 2):
+            for k in range(-1, 2):
+                row_indexer = min(max(row + k, 0), dem.shape[0] - 1)
+                col_indexer = min(max(col + j, 0), dem.shape[1] - 1)
+                Z[count] = dem[row_indexer, col_indexer]
+                count += 1
+
+        # Get a mask of all invalid (nan or inf) values.
+        invalids = ~np.isfinite(Z)
+
+        # Skip the pixel if it and all of its neighbours are invalid
```

Comment:
This is something we had no control over before, choosing the behaviour with NaNs will be really cool

Reply:
What kind of options would you like?

Reply:
I guess we could have:
`default` = return NaN if any of the 9 values are NaN (default in GDAL, SAGA)
`fill_mean` = return value using the mean of the 9 values to fill gaps, if center pixel is not a NaN
`fill_median` = return value using the median of the 9 values to fill gaps, if center pixel is not a NaN
`fill_bilinear` = return value after bilinear extrapolation of the 3x3 data (if possible)
`fill_cubic` = return value after cubic extrapolation of the 3x3 data (if possible)

Seems to me that the extrapolation will fail quite often if there is a lot of NaN, would it make sense to have a similar extrapolation but performed on a slightly larger kernel (5x5)? I'm also wondering if this would work better directly in the numba loop, or with a pre-filled DEM at specific locations (e.g. using a `binary.dilate(niter=2)` on a mask of valid data).
What do you think?

Reply:
I can definitely implement the first three filling cases, and I'll try the other.

Changing kernel sizes is a bit harder, as the calculation for A, B, C... is based on 9 numbers and not more. I'm convinced there's a way to calculate them on more numbers, but that's way beyond my understanding of this!

Reply:
> Is there a typo in planform/profile curvature here, @rhugonnet? They seem to be the same only with an inverted sign.

Yes, my bad, corrected versions:
`PROFC = -2(D*G² + E*H² + FGH)/(G² + H²)`
`PLANC = 2(D*H² + E*G² - FGH)/(G² + H²)`

Equations 17 and 18 of Zevenbergen, 1987.

Reply:
I added `"none"`, `"mean"` and `"median"` in e940f80. For the rest, I'd prefer to raise an issue for the future!